examples/labels.py

```diff
@@ -0,0 +1,97 @@
+            # see:
+            # - https://www.geeksforgeeks.org/python-exception-handling/
+            # - https://docs.python.org/3/tutorial/errors.html#raising-exceptions
+
+        self.label = label
+        self.__rule = rule
+
+    # This is a wrapper method around the private rule attribute we
+    # store on self during init.
+    #
+    # Q: Why make the rule private and
+    # wrap the call to it in another method?
+    # A: This encapsulation allows us to expose a nicer set of behavior
+    # and naming conventions to both the user and ourselves as developers.
+    # In `label_features` below, you see that we can then check whether
+    # a log `matches` the definition which reads more like plain english
+    # and is an important part of writing clean, idiomatic python code.
+    # TODO: Implement this wrapper function by using the _rule attribute
+    def matches(self, log: Dict[str, Any]) -> bool:
+        if log == self.__rule:
+            return log
+
+
+def label_features(
+    logs: List[Dict[str, Any]], definitions: List[FeatureDefinition]
+) -> List[Dict[str, Any]]:
+    # Iterate through all the logs
+    for log in logs:
+        # Check whether the log matches the definition
+        # for each definition supplied in the defintions list
+        for definition in definitions:
+            # NOTE: This reads much like an English sentence
+            # and is self-explanatory. I don't need to read the
+            # implementation logic to get a sense of what's happening
+            if definition.matches(log):
+                # NOTE: Since we're mutating the log itself and interacting
+                # with a field that may (does) not already exists, we need
+                # to first check if it is present in our log and instantiate
+                # it if not.
+                if "labels" not in log:
+                    log.update({"labels": list()})
+                log["labels"].append(definition.label)
+    return logs
+
+
+###########################################################
+# Example of how the FeatureDefintion class works
+#
+# The following if __name__ == "__main__" syntax
+# is a way to tell python that if your run this file
+# as a script from the command line, then this is the code
+# that needs to be executed.
+###########################################################
+if __name__ == "__main__":
+    # TODO: Import logs from JSON file here ...
+    json_file = input("Enter your json file path: ")
+    logs = pd.read_json(json_file)
+    
+
+    # Create a map rule to test out the FeatureDefinition with
+    def map_rule(log: Dict[str, Any]) -> bool:
+        return "pageUrl" in log and "map" in log["pageUrl"]
+
+    def container_rule(log: Dict[str, Any]) -> bool:
+        return "path" in log and "container" in log["path"]
+    
+    def table_rule(log: Dict[str, Any]) -> bool:
+        return "path" in log and "table" in log["path"]
```

Comment:
Good start. It's also good practice to think about rules that will _break_ your code when testing. In short, we want to try to understand that our code works correctly under all the conditions we expect it to work correctly. Also, it should not only break under the conditions we expect it to break, but break in the ways we want it to (like raising specific errors).

You've created two new code objects that require testing here:
1. `FeatureDefinition` class
2. `label_features` function

So far you've come up with some example rules, all of which should work. Now try coming up with some rules that _shouldn't_ work (i.e. FeatureDefintion class will either work improperly OR throw the TypeError you specificed).

Also, consider different test cases for your label_features function. Right now you've passed in values that match what it expects. Try passing it values it doesn't expect, like different variable types, things that aren't logs, etc. You'll get errors back, so think about ways you can catch issues earlier in the function and return more informative errors to the user, similar to how we do in the FeatureDefinition __init__ method.